File: Week_01/G20190343020108/189-rotate-array.js
/** 解法一：暴力法
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function(nums, k) {
    const size = nums.length;
    if (size < 2) return;
    const offset = k % size;
    for (let i = 0; i < offset; i++) {
        // 外层循环记录移动趟数
        // 内层循环数组群移
        let temp = nums[size - 1];
        for (let j = size - 1; j > 0; j--) {
            nums[j] = nums[j - 1];
        }
        nums[0] = temp;
    }
};

/** 解法二：倒序法
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate2 = function(nums, k) {
    const size = nums.length;
    if (size < 2) return;
    const offset = k % size;
    nums.reverse(); // 1. 反转整个数组
    reversePartial(nums, 0, offset - 1); // 2. 反转左半部分数组
    reversePartial(nums, offset, size - 1); // 2. 反转右半部分数组
};

// 反转数组中指定区间的元素
const reversePartial = (arr, start, end) => {
    while (start < end) {
        [arr[start], arr[end]] = [arr[end], arr[start]];
        start++;
        end--;
    }
};

module.exports = { rotate, rotate2 };

File: Week_01/G20190343020350/week1/189-rotate-array.cpp
class Solution {
public:
    void reverseArray(vector<int>& nums, int start, int end) {
        int tmp = 0;
        while(start < end) {
            tmp = nums[end];
            nums[end] = nums[start];
            nums[start] = tmp;
            end--;
            start++;
        }
    }
    void rotate(vector<int>& nums, int k) {
        k %= nums.size();
        reverseArray(nums, 0, nums.size()-1);
        reverseArray(nums, 0, k-1);
        reverseArray(nums, k, nums.size()-1);
    }
};
